Dpinger   with selectable use of apinger or dpinger

marjohn56 opened the pull request:
This version of Dpinger allows the user to select whether to use dpinger or apinger, the setting is in System->General->Settings.

 I have tried wherever possible to keep functions seperate, thus retaining as much of the existing flow as possible. Alarms checked and working.  Some  advanced settings are now hidden when dpinger is selected.

@Franco can you look at deleting the RRD files whenever the system flips from apinger to dpinger or vice versa, I searched but could not find that.

marjohn56 commented:
Update 1 - Added rc.gateway-alarm script. We can use it without, but we lose logging info.
Update 2 - Added fix for monit, it does no like ':' in the pid filenames, so str_replace to '-',

fichtner commented on src/etc/inc/plugins.inc.d/core.inc:
```diff
@@ -82,16 +82,48 @@ function core_services()
-        $pconfig['pidfile'] = '/var/run/apinger.pid';
-        $services[] = $pconfig;
-    }
+    if (!isset($config['system']['prefer_dpinger'])) {
```
if you push this to is_apinger_enabled and is_dpinger_enabled this doesn't have to be evaluated here and dpinger plugs in seamlessly here

marjohn56 replied:
just belts and braces... but I'll remove it if you think it's unnecessary.

marjohn56 replied:
I've moved the catches over to the services...

fichtner commented on src/etc/inc/plugins.inc.d/core.inc:
```diff
@@ -82,16 +82,48 @@ function core_services()
+            $pattern = "/^dpinger_({$name})~([^~]+)~([^~]+)\.pid$/";
+            foreach ($pidfiles as $pidfile) {
+                if (preg_match($pattern,basename($pidfile), $matches)) {
+                    $dpinger_pid = $pidfile;
```
wouldn't it be enough to construct the pid file here instead of glob and find ?

this loop must also take into account monitoring_disabled flag and skip that particular dpinger process because it doesn't exist

marjohn56 replied:
The pid filename is created in gwlib and the name is horribly complex. At first I could see no reason for the name being as complex as it is... but after thinking about it, I can see the reason behind it,.

fichtner replied:
which is? :)

marjohn56 replied:
You may have multiple targets on the same interface.

fichtner replied:
but we only have one interface to monitor?

marjohn56 replied:
I suppose because the target is the variable?  best ask Denny on that one. :)

fichtner commented on src/etc/inc/plugins.inc.d/core.inc:
```diff
@@ -82,16 +82,48 @@ function core_services()
+            }
+            $pconfig = array();
+            $pconfig['name'] = "dpinger {$gateway['name']}";
+            $pconfig['description'] = gettext("Gateway Monitoring Daemon");
```
We need to embed the gateway name here, e.g. sprintf(gettext('Gateway Monitor %s'), $the_gateway_name)

marjohn56 replied:
Currently looks like this:

![image](https://user-images.githubusercontent.com/17962918/38539200-fddd4b38-3c8e-11e8-8ae2-f530db535699.png)

fichtner commented on src/etc/inc/plugins.inc.d/core.inc:
```diff
@@ -82,16 +82,48 @@ function core_services()
+                }
+            }
+            $pconfig = array();
+            $pconfig['name'] = "dpinger {$gateway['name']}";
```
and just 'dpinger' here

marjohn56 replied:
No, there are multiple dpinger instances.

fichtner commented on src/www/system_gateways_edit.php:
```diff
@@ -764,6 +767,7 @@ function calculated_change() {
                     </div>
                   </td>
                 </tr>
+<?php           if (!isset($config['system']['prefer_dpinger'])):?>
```
this looks very nice, thanks ❤️

fichtner commented on src/www/system_general.php:
```diff
@@ -46,6 +46,7 @@
     $pconfig['hostname'] = $config['system']['hostname'];
     $pconfig['language'] = $config['system']['language'];
     $pconfig['prefer_ipv4'] = isset($config['system']['prefer_ipv4']);
+    $pconfig['prefer_dpinger'] = isset($config['system']['prefer_dpinger']);
```
note to self: should be in misc, along with the default gateway switching option which is currently under firewall

(it's not a basic setting)

fichtner commented on src/etc/rc.gateway_alarm:
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+#
```
this file is better off with a copyright block and your copyright line

marjohn56 replied:
It's a rip of the pf* file, I cannot really change it to copyright opnsense... :)

fichtner replied:
it's a problem, because if they don't assert a license this is not open source licensed code tainting their product

then again, this isn't even called anywhere?

marjohn56 replied:
line 1421 in gwlib.

fichtner replied:
ok, here's a clean version based on our code. you can use it as is: /usr/local/etc/rc.monitor

fichtner replied:
(https://github.com/opnsense/core/commit/6fb26f989)

marjohn56 replied:
I could, but I would lose the alarm logging info wouldn't I?

fichtner replied:
refactored what we have in apinger to prepare for your required changes so you plug the logging into it

fichtner commented on src/etc/inc/services.inc:
```diff
@@ -1897,3 +1896,26 @@ function service_control_restart($name, $extras)
+    if (isset($config['gateways']['gateway_item'])) {
+        foreach($config['gateways']['gateway_item'] as $gwkey => $gateway) {
+            if (!isset($gateway['monitor_disable']) || $gateway['monitor_disable'] == "0") {
+                if (isset($config['system']['prefer_dpinger'])) {
```
ah, already here... nice, but you can move this to the top :)

fichtner commented on src/www/widgets/widgets/gateways.widget.php:
```diff
@@ -78,6 +80,7 @@ function gateways_widget_update(sender, data)
         <tr>
             <th><?=gettext('Name')?></th>
             <th><?=gettext('RTT')?></th>
+            <th><?=gettext('Dev')?></th>
```
Let's use STDEV consistently... https://www.abbreviations.com/abbreviation/Standard+Deviation

marjohn56 replied:
Still think it should be RTTd :)

fichtner commented on src/www/widgets/api/plugins/gateway.inc:
```diff
@@ -49,6 +50,9 @@ function gateway_api()
                     $gatewayItem['status_translated'] = gettext("Latency");
                     break;
+                case "stdev":
+                    $gatewayItem['status_translated'] = gettext("StDev");
```
STDEV

marjohn56 replied:
Actually it should be RTTd, - That changed in V2 but fell through my review.

fichtner replied:
Fine, it should simply be the same everywhere

fabianfrz replied:
Or "Standard Derivation"

marjohn56 replied:
Not Derivation, that's from derived.

Deviation :)

fabianfrz replied:
oops

fichtner commented on src/www/system_gateways.php:
```diff
@@ -297,6 +297,7 @@ function delete_gateway_item($id, $a_gateways)
                     <th class="hidden-xs hidden-sm hidden-md"><?=gettext("Gateway"); ?></th>
                     <th class="hidden-xs hidden-sm hidden-md"><?=gettext("Monitor IP"); ?></th>
                     <th class="text-nowrap hidden-xs"><?=gettext("RTT"); ?></th>
+                    <th class="text-nowrap hidden-xs"><?=gettext("RTTd"); ?></th>
```
STDEV

fichtner commented on src/etc/inc/gwlb.inc:
```diff
@@ -52,13 +52,20 @@ function setup_gateways_monitor($verbose = false)
         echo 'Setting up gateway monitors...';
```
this file should hold your copyright, too

marjohn56 commented on src/etc/inc/interfaces.inc:
```diff
@@ -4769,3 +4769,20 @@ function get_carp_interface_status($carpinterface)
     }
     return;
 }
+function interface_wait_tentative($interface, $timeout = 10) {
```
OK, I'll test it without and see if there is any difference, Isn't it better to have something rather than nothing though?

fichtner replied:
it just waits here trying to avoid the error. I want to see the error first... there should be a better way to handle this

marjohn56 commented on src/etc/inc/gwlb.inc:
```diff
@@ -1034,3 +1084,411 @@ function validate_address_family($ipaddr, $gwname)
+			return (false);
+		}
+	}
+	return mwexec("/usr/bin/dpinger {$params} {$gateway['monitor']} >/dev/null");
```
Okey dokey.

fichtner replied:
this :)

marjohn56 commented:
OK.. I think I have it all in place now. Confused on the gateway-alarm bit, are we calling gateway alarm or rc.monitor... or do you want me to take rc.monitor and use that as a template for rc.gateway-alarm?

fichtner commented:
rc.monitor is a common file for both apinger and dpinger... it mainly reloads the proper services in the right order (there was a fix here recently so we should keep that code in one place for maintenance and clarity)... and so if dpinger is used, which sets a second argument the log line can be added in full

marjohn56 commented:
OK..  try that.. I have not touched rc.monitor at this point. Needs the catch for the logging etc.

I have added a little bit in system_general in an attempt to delete the rrd files when the user changes between apinger <> dpinger.

marjohn56 commented:
f*&&^&*!!!!!  GIT it's done me again!!

I'll have to rebuild the PR from my test VM

fabianfrz commented:
`git reset --hard head^` may do it (removes the last commit)

marjohn56 commented:
Would be nice to think so, but sadly not. It only ever happens with Windows GIT and shell... says it all really.

Fortunately my VM has all the files on there as I copy and test them before a PR, I also copied over the files after modifying them this morning so they are still there. I am making some changes @fichtner wanted to the placement of the config entry and then I'll  push it again..

marjohn56 commented:
OK... back to normal. ;)

Modified rc.monitor a little
Added mwexec call to rm *quality.rrd files on change
moved the config to system->advanced-misc

Think that's it... until you find something else. :)

fichtner commented on src/www/system_advanced_misc.php:
```diff
@@ -221,8 +228,8 @@ function thermal_modules()
     }
 ?>
       <section class="col-xs-12">
-        <form method="post" name="iform" id="iform">
```
looks like this clashes with a newer commit on master, reverting the style update

marjohn56 commented:
Try that.

marjohn56 commented:
@fichtner 

There is a problem if you change the name of the service to just dpinger in core,inc like so:

$pconfig = array();
            $pconfig['name'] = "dpinger";
            $pconfig['description'] = sprintf(gettext('Gateway Monitor %s'), $gateway['name']);
            $pconfig['php']['restart'] = array('setup_gateways_monitor');
            $pconfig['php']['start'] = array('setup_gateways_monitor');
            $pconfig['pidfile'] = $dpinger_pid;
            $services[] = $pconfig;    
            unset($dpinger_pid);

What happens  then is that in the services widget, you have two services called dpinger, and you cannot stop or restart them properly. Afraid it is going to have to stay the way it is, unless you can think of another way,

fichtner commented:
Solved this for OpenVPN a good while back. I'll take care of it in a manual merge. It needs an additional ID parameter... or the gateway name or so. Easy. :)

marjohn56 commented:
Okay Obi-Wan,  I'll leave you to play with that one. :)

marjohn56 commented:
When you hit start on either one it restarts both anyway, that can be lived with stopping it's not quite so good,

fichtner commented on src/etc/rc.monitor:
```diff
@@ -37,3 +37,13 @@ fi
+ALARM_RTTSD="${5}"
+ALARM_LOSS="${6}"
+
+if [ -z "$#" -eq 6 ]; then
```
-z only takes one argument and "sh -n src/etc/rc.monitor" should return a syntax error... best to compare a string here right away: [ "${#}" = "6" ]

marjohn56 replied:
How about if [ "$#" -eq 6 ]

fichtner replied:
works too, practically the same as the string match minus the certainty of how "$#" is converted into a number... :)

fichtner commented on src/etc/rc.monitor:
```diff
@@ -37,3 +37,13 @@ fi
+ALARM_LOSS="${6}"
+
+if [ -z "$#" -eq 6 ]; then
+    echo ">>> Gateway alarm: ${GW} (Addr:${ALARM_ADDR} Alarm:${ALARM_FLAG} RTT:${ALARM_RTT}ms STDEV:${ALARM_RTTSD}ms Loss:${ALARM_LOSS}%)" | /usr/bin/logger -p daemon.info -i -t rc.monitor
```
It's ${GATEWAY} in this script

marjohn56 commented:
So what I have done will work? I find bourne shell a bit strange,

fichtner commented:
Yep, will merge this till Friday. Currently without proper Internet and therefore test opportunity -.-

marjohn56 commented:
OK, well I won't be around after tomorrow night for a couple of weeks+ so good luck. :)

fichtner commented:
alright... enjoy the time off... :)

I'll push out the dhcp6c switch and the associated configuration code rework in 18.1.7. It's ready for release now that 18.1.6 is behind us.


Cheers,
Franco

marjohn56 commented:
RRD Fix

fichtner commented:
Closing as discussed, mostly manually merged so several conflicts in here.
